File: services/ai-agents/app/presentation/http/result_mapper.py
from __future__ import annotations

from fastapi import HTTPException

from ...domain.shared.exceptions import (
    ConflictException,
    DomainException,
    InvariantViolation,
    NotFoundException,
    UnauthenticatedException,
    ValidationException,
)


def _status_for(err: DomainException) -> int:
    if isinstance(err, ValidationException):
        return 400
    if isinstance(err, InvariantViolation):
        return 400
    if isinstance(err, UnauthenticatedException):
        return 401
    if isinstance(err, NotFoundException):
        return 404
    if isinstance(err, ConflictException):
        return 409
    return 500


def raise_for_error(err: DomainException) -> None:
    """Translate a domain error into an HTTP exception. Controllers call
    this so they don't repeat status-code switches."""
    raise HTTPException(
        status_code=_status_for(err),
        detail={"error": err.code, "message": err.message},
    )

File: services/auth/src/presentation/http/resultMapper.ts
import { Response } from "express";
import { Result } from "../../domain/shared/Result";
import {
  ConflictException,
  DomainException,
  NotFoundException,
  UnauthenticatedException,
  ValidationException,
  InvariantViolation,
} from "../../domain/shared/DomainException";

// Central place where domain errors become HTTP status codes. Adding a new
// DomainException only requires one new branch here — controllers stay clean.
export function respondWithResult<T>(
  res: Response,
  result: Result<T, DomainException>,
  successStatus = 200,
): void {
  if (result.isOk()) {
    res.status(successStatus).json(result.getValue());
    return;
  }
  const err = result.getError();
  const status = mapStatus(err);
  res.status(status).json({ error: err.code, message: err.message });
}

function mapStatus(err: DomainException): number {
  if (err instanceof ValidationException) return 400;
  if (err instanceof InvariantViolation) return 400;
  if (err instanceof UnauthenticatedException) return 401;
  if (err instanceof NotFoundException) return 404;
  if (err instanceof ConflictException) return 409;
  return 500;
}
